src/noteburst/worker/main.py

```diff
@@ -86,56 +77,68 @@ async def startup(ctx: dict[Any, Any]) -> None:
         )
         ctx["slack"] = slack_client
 
-    jupyter_image: nc_models.NubladoImage | None = None
-    if config.image_selector == "reference":
-        jupyter_image = nc_models.NubladoImageByReference(
-            reference=config.image_reference
-        )
-    elif config.image_selector == "weekly":
-        jupyter_image = nc_models.NubladoImageByClass(
-            image_class=nc_models.NubladoImageClass.LATEST_WEEKLY
-        )
-    else:
-        # "Recommended" is default
-        jupyter_image = nc_models.NubladoImageByClass()
-
+    # Seed an initially-available bot user identity. The while loop below
+    # will acquire new identities if spawning with this one fails.
+    identity_manager = IdentityManager.from_config(config)
+    ctx["identity_manager"] = identity_manager
     identity = await identity_manager.get_identity()
 
+    # Loop with different identities until we get a successful spawn
+    nublado_pod: NubladoPod | None = None
+    spawn_exception: Exception | None = None
     while True:
-        logger = logger.bind(worker_username=identity.username)
-
-        jupyter_client = NubladoClient(
-            user=await _get_client_user(identity, config, http_client, logger),
-            base_url=str(config.environment_url),
-            logger=logger,
-            hub_route=config.jupyterhub_path_prefix,
-        )
-
-        await jupyter_client.auth_to_hub()
         try:
-            await jupyter_client.spawn_lab(config=jupyter_image)
-            if config.image_reference:
-                logger = logger.bind(image_ref=config.image_reference)
-            else:
-                logger = logger.bind(image_ref=config.image_selector)
-            # We don't currently expose the reference of the actually-spawned
-            # image in the client, so put that down as a to-do item.
-            async for _ in jupyter_client.watch_spawn_progress():
-                continue
-            await jupyter_client.auth_to_lab()
+            nublado_pod = await NubladoPod.spawn(
+                identity=identity,
+                nublado_image=config.nublado_image,
+                http_client=http_client,
+                user_token_scopes=config.parsed_worker_token_scopes,
+                user_token_lifetime=config.worker_token_lifetime,
+                logger=logger,
+            )
             break
-        except JupyterProtocolError as e:
+        except (
+            JupyterProtocolError,
+            # Happens when we have orphaned pods and internal jupyterhub errors
+            JupyterWebError,
+            JupyterSpawnError,
+            JupyterTimeoutError,
+            JupyterWebSocketError,
+            # From the User login with Gafaelfawr
+            httpx.HTTPError,
+        ) as e:
             logger.warning("Error spawning pod, will re-try with new identity")
```

Comment:
Why shunt the exception info to the debug level here? I think you can include it in the warning message with exc_info=e